src/Microsoft.ML.Data/TrainCatalog.cs

```diff
@@ -674,6 +674,31 @@ internal RankingTrainers(RankingCatalog catalog)
             var eval = new RankingEvaluator(Environment, options ?? new RankingEvaluatorOptions() { });
             return eval.Evaluate(data, labelColumnName, rowGroupColumnName, scoreColumnName);
         }
+
+        /// <summary>
+        /// Run cross-validation over <paramref name="numberOfFolds"/> folds of <paramref name="data"/>, by fitting <paramref name="estimator"/>,
+        /// and respecting <paramref name="samplingKeyColumnName"/> if provided.
+        /// Then evaluate each sub-model against <paramref name="labelColumnName"/> and return metrics.
+        /// </summary>
+        /// <param name="data">The data to run cross-validation on.</param>
+        /// <param name="estimator">The estimator to fit.</param>
+        /// <param name="numberOfFolds">Number of cross-validation folds.</param>
+        /// <param name="labelColumnName">The label column (for evaluation).</param>
+        /// <param name="samplingKeyColumnName">Name of a column to use for grouping rows. If two examples share the same value of the <paramref name="samplingKeyColumnName"/>,
+        /// they are guaranteed to appear in the same subset (train or test). This can be used to ensure no label leakage from the train to the test set.
+        /// If <see langword="null"/> no row grouping will be performed.</param>
+        /// <param name="rowGroupColumnName">The name of the groupId column in <paramref name="data"/>.</param>
+        /// <param name="seed">Seed for the random number generator used to select rows for cross-validation folds.</param>
+        /// <returns>Per-fold results: metrics, models, scored datasets.</returns>
+        public IReadOnlyList<CrossValidationResult<RankingMetrics>> CrossValidate(
+            IDataView data, IEstimator<ITransformer> estimator, int numberOfFolds = 5, string labelColumnName = DefaultColumnNames.Label,
+            string samplingKeyColumnName = DefaultColumnNames.GroupId, string rowGroupColumnName = DefaultColumnNames.GroupId, int ? seed = null)
```

Comment:
Review: Should samplingKeyColumnName be interchangeable with rowGroupColumnName, meaning the user should just provide one of them?

Reply:
The way I understand it, which to be honest isn't 100% clear, `samplingKeyColumnName` and `rowGroupColumnName` should always be the same with the implementation you're adding (i.e. the `samplingKeyColumn` you're passing to `CrossValidateTrain` should be the same as the `rowGroupColumnName passed to the `Evaluate` method), because you'd want to group the samples when evaluating in the same way that they were grouped during training (else the "score" predicted by the model wouldn't make sense)... right? or not?

So my personal opinion on this regard is that we should drop one of the two parameters, let's say drop `rowGroupColumnName` and just stick with `samplingKeyColumn`. And pass `samplingKeyColumn` to both `CrossValidateTrain` and `rowGroupColumnName`. If in the future we find that there's actually use on having 2 different parameters for grouping data on these 2 stages (training vs evaluation), then we'll simply add another overload.

The only thing that confuses me is this comment left by a former member of ML.NET: [https://github.com/dotnet/machinelearning/issues/2685#issuecomment-466171787](https://github.com/dotnet/machinelearning/issues/2685#issuecomment-466171787) . I don't fully understand the example he's giving about ranking author's sentences... but It does seem to imply that we should have 2 different parameters for grouping data, _and_ also that we should change how CrossValidation is done for Ranking in order to support those 2 parameters. Still, it's unclear to me if this is a _must_ , or if it was simply a feature idea... (if it was just a simple feature idea, we could get away with not having to change CV right away, and just stick with the CV with only 1 `samplingKeyColumnName` parameter.

---
In reply to: [446441698](https://github.com/dotnet/machinelearning/pull/5263#discussion_r446441698) [](ancestors = 446441698)

Reply:
Ok. After thinking about it, I do think that we need 2 parameters for grouping: one for splitting the data and the other one for training _and_ evaluation, and these are only needed for ranking. This will also need some changes in the `CrossValidateTrain` method  in order to use those 2 parameters, that are to be provided _only_ for ranking, but I wouldn't expect them to be major changes.

If I'm right about this, then I think that the way CV is run for ranking through NimbusML and MAML is just plainly wrong at the moment, and also that my understanding of #3711 has changed.

---
In reply to: [446457287](https://github.com/dotnet/machinelearning/pull/5263#discussion_r446457287) [](ancestors = 446457287,446441698)

Reply:
For the record, after having some discussions on this topic, we've decided to keep only the `rowGroupColumnName` parameter and use it internally both as the samplingKeyColumn (when splitting the data) and groupIdColumn (when training and evaluating the ranking models); this to ensure that samples with the same GroupId always belong to the same fold.

After the next release we'll explore adding another API with the `samplingKeyColumn` parameter to provide another level of grouping when splitting the data, while still using the `rowGroupColumnName` parameter for training and evaluating. It's useful to have 2 separate APIs to not confuse the users.

---
In reply to: [446465102](https://github.com/dotnet/machinelearning/pull/5263#discussion_r446465102) [](ancestors = 446465102,446457287,446441698)